### ISSUE TITLE
Bug executing`pvget` on unchanged PV in remote server.

### ISSUE THREAD

**Issue by jembishop** [epics-base/pvaPy#85]
Hello. 

I am encountering an issue with running a `pvaPy` server. To replicate the bug one must:

- Have a `pvaPy` server running on a remote machine. 
- Execute `pvget` on a PV. 
- Before the PV changes, execute another `pvget`. 

In this circumstances we are seeing the second value returned by `pvget` is always a "zero" struct. Inspection of the traffic using wireguard finds that the body of the message is `00`. 
I initially raised this [here](https://github.com/mdavidsaver/p4p/issues/124)  but I thought it was a bug with the `PVAGateway` in `p4p` at that point. I can confirm this issue only occurs with the `pvaPy` implementation of the server. mdavidsaver did in fact fix an issue which looks like it could cause this, but it turned out to not be applicable in my case. 

Thanks, Jem

**Comment by sveseli** [comment]
Hi,

Just to make sure I understand, you execute second pvget before the first one returns?
Which version of pvapy are you using, and on which OS are you running? What kind of record are you serving? Perhaps a simplified version of your server code would help debug this. It would also help if you describe a bit your use case, what do clients do, how many clients are there, what are they written in, etc.


Thanks,

Sinisa

**Comment by jembishop** [comment]
Hi Sinisa, thanks for getting back to me.

Our installed packages:
```
(test-pvapy-py3.10) ➜  test-pvapy git:(main) ✗ pip freeze  
numpy==1.25.2
pvapy==5.3.0
```

The OS info is:

```
root@c9446aaf738c:/app# cat /etc/os-release 
PRETTY_NAME="Debian GNU/Linux 12 (bookworm)"
NAME="Debian GNU/Linux"
VERSION_ID="12"
VERSION="12 (bookworm)"
VERSION_CODENAME=bookworm
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```
The minimal server script is:

```
from pvaccess import PvaServer, INT, FLOAT, PvObject
import time
import random
import logging

logging.basicConfig(level=logging.DEBUG)

pv = PvObject({'value' : FLOAT})


pvaServer = PvaServer('hello', pv)  # Initializes server with first record

while True:
    pv["value"] += random.random() - 0.5
    print(pv["value"])
    time.sleep(5)
```

The minimal client script is:
```
import os
from p4p.client.thread import Context
from datetime import datetime
import time
import logging
logging.basicConfig(level=logging.DEBUG)
c = Context("pva")
print(datetime.now())
pv = "hello"
v = c.get(pv)
print(datetime.now(), v)
time.sleep(0.1)
print(datetime.now())
v = c.get(pv)
print(datetime.now(), v)
c.close()
```

However if you repeat`pvget hello` in rapid succession you will also encounter the same bug, so it's not an issue with `p4p`. 
Note there is no concurrency here, its all synchronous I am just running this script once not at the same time in two different terminals or something. The result of the first request always arrives before I start the second. The wireshark traffic is showing the same thing, the two requests are not overlapping the packets. 

Our use case in production is running a server which converts Vista/VSystem PVs into EPICS PVs, which is where we are also encountering this issue. 

I would be interested if you can replicate this as well. I am only able to reproduce this when the client and server are running on different hosts (running two docker containers on the same host doesn't seem to reproduce, neither does running the server in a docker container, and a client on the host on the same machine). 

Thanks, Jem

**Comment by sveseli** [comment]
Hi Jem,

Could you please try the small changes to your server code shown below and let me know if you still see the issue?

<pre>
from pvaccess import PvaServer, INT, FLOAT, PvObject
import time
import random
import logging

logging.basicConfig(level=logging.DEBUG)

pv = PvObject({'value' : FLOAT})


# Forces using atomic record updates
pvaServer = PvaServer('hello', pv.copy())

while True:
    pv["value"] += random.random() - 0.5
    print(pv["value"])
    # Atomic record update
    pvaServer.update(pv)
    time.sleep(5)
</pre>

Thanks.

Sinisa

**Comment by jembishop** [comment]
Thanks for your suggestion Sinisa, unfortunately this doesn't seem to change anything. 
After some more experimentation, it appears I don't seem to get the bug running the server/client between two other hosts. 
This perhaps has some interaction with how containers in the docker network in our docker compose setup are connected to hosts outside then.

**Comment by sveseli** [comment]
I cannot replicate the issue in my environment using the server code above and multiple pvget commands. If your docker compose setup were easily reproducible, I could try running it myself. The server code you had originally would be problematic if your record had a more complex PV that had multiple fields, etc., but the problem that you would be seeing would be different (not all fields updated at the same time). I have never seen an issue with "zero" struct that you describe. Is the problem intermittent, i.e., do you actually get data after one of the requests fails, or do all requests continue failing after the first one fails?

Could you perhaps instead of using subsequent get requests try running monitor in your client for a while, and see if you notice any issues getting the updates?

**Comment by jembishop** [comment]
The issue seems to replicate 100% of the time, when a PV is changed, so not intermittant. A monitor works completely fine, it does not give any zero values (which makes sense as it only listens for changes).

**Comment by sveseli** [comment]
Hmm, so you get a value okay the first time you issue pvget, then subsequent pvget requests return empty structure. What happens when the record changes? Do you get the update at least once?

You mentioned that you only see this problem when using containers. The two hosts where you do not see the issue, are those regular machines?

Sorry for all these questions, I am trying to figure out if I could reproduce this somehow. Perhaps if your container images were available, I could run them locally and see what is going on.

Thanks,

Sinisa

**Comment by jembishop** [comment]
Yes we do see the updated value. The issue is that if say the pv on the server is 0.5 and changes to 1.5 after 10 seconds, initially the pv get will return 0.5. Then after 5 seconds, if you do any number of pvgets it will return 0. Then when you pass 10 seconds when the value on the server is changed, a pvget will return the now changed value of 1.5. 

The issue is that middle requests should not return 0, they should return 0.5.

the machines in question which I tried to replicate it on, but don’t have the bug are regular machines, not virtual machines or something like that.

I am trying to get a setup together which should help you replicate, it’s proving quite difficult but hopefully I can get something for you. Thank you for your continuing help, ask as many questions as you want, I am almost out of ideas with this one! If there is any way to turn on some more detailed logging that would ideal too.

**Comment by sveseli** [comment]
Logging is a good idea, I should have thought of it before. There are two variables that control logging: PVAPY_EPICS_LOG_LEVEL (for epics libraries) and PVAPY_LOG_LEVEL (pvapy code). In this case, the problem will most likely be in the underlying epics C++ libraries, so try running your server using something like this: 

<pre>
$ PVAPY_EPICS_LOG_LEVEL=0 python server.py
</pre>

You should see messages like this:

<pre>
2023-09-28T09:11:37.427 UDP socket 0.0.0.0:0 closed.
2023-09-28T09:11:37.427 UDP Server Rx (0) 0.0.0.0:52204 <- <Ukn Addr Type>
2023-09-28T09:11:37.427 Thread 'UDP-rx 0.0.0.0:0' exiting.
2023-09-28T09:11:37.427 UDP socket 164.54.29.86:5076 closed.
2023-09-28T09:11:37.428 UDP Server Rx (0) 164.54.29.86:5076 <- 164.54.49.88:42364
2023-09-28T09:11:37.428 Thread 'UDP-rx 164.54.29.86:5076' exiting.
2023-09-28T09:11:37.428 UDP socket 164.54.29.255:5076 closed.
2023-09-28T09:11:37.428 UDP Server Rx (0) 164.54.29.255:5076 <- 164.54.29.86:52204
2023-09-28T09:11:37.428 Thread 'UDP-rx 164.54.29.255:5076' exiting.
2023-09-28T09:11:37.428 UDP socket 224.0.0.128:5076 closed.
2023-09-28T09:11:37.428 UDP Server Rx (0) 224.0.0.128:5076 <- 164.54.29.86:5076
2023-09-28T09:11:37.428 Thread 'UDP-rx 224.0.0.128:5076' exiting.
2023-09-28T09:11:37.428 Stopped accepting connections at 0.0.0.0:5075.
</pre>

Let me know what comes out of this.

**Comment by mdavidsaver** [comment]
> The issue is that middle requests should not return 0, they should return 0.5.

This issue might be consistent with a BitSet being incorrectly shared between GET operations.  A server side update sets some bits, then the first GET operation clears these bits.  The bits then remain cleared until the next server update sets them.

**Comment by jembishop** [comment]
Thanks, sevesli and mdavidsaver. I see I made one important error in my previous tests, in the container in prod we were on python 3.6, and therefore the newly installed pvapy was at an older version. Using the latest python3.11 I am now on `pvapy==5.3.0` . However this doesn't solve the bug. Running the command with more detailed logging gives me
```
[root@base base]# PVAPY_EPICS_LOG_LEVEL=0 /root/.pyenv/shims/python3.11 server.py | grep -v "UDP"
2023-09-28T17:44:29.673 Creating acceptor to 0.0.0.0:5075.
2023-09-28T17:44:29.673 Socket bind error: Address already in use.
2023-09-28T17:44:29.673 Configured TCP port 5075 is unavailable, trying to assign it dynamically.
2023-09-28T17:44:29.673 Creating acceptor to 0.0.0.0:5075.
2023-09-28T17:44:29.673 Using dynamically assigned TCP port 38207.
2023-09-28T17:44:29.673 Accepting connections at 0.0.0.0:38207.
2023-09-28T17:44:29.673 Creating datagram socket from: 0.0.0.0:56002.
2023-09-28T17:44:29.673 Broadcast address #0: 172.20.255.255:5076. (not unicast)
2023-09-28T17:44:29.673 Broadcast address #1: 172.18.255.255:5076. (not unicast)
2023-09-28T17:44:29.674 Creating datagram socket from: 172.20.11.45:5076.
2023-09-28T17:44:29.674 Creating datagram socket from: 172.20.255.255:5076.
2023-09-28T17:44:29.674 Creating datagram socket from: 172.18.0.21:5076.
2023-09-28T17:44:29.674 Creating datagram socket from: 172.18.255.255:5076.
2023-09-28T17:44:29.674 Creating datagram socket from: 224.0.0.128:5076.
2023-09-28T17:44:29.674 Local multicast enabled on 127.0.0.1/224.0.0.128:5076.
2023-09-28T17:44:29.674 Sending 47 bytes 0.0.0.0:56002 -> 172.20.255.255:5076.
2023-09-28T17:44:29.674 Sending 47 bytes 0.0.0.0:56002 -> 172.18.255.255:5076.
2023-09-28T17:44:44.675 Sending 47 bytes 0.0.0.0:56002 -> 172.20.255.255:5076.
2023-09-28T17:44:44.675 Sending 47 bytes 0.0.0.0:56002 -> 172.18.255.255:5076.
```

Which seems similar to when I run it on my test server (not in a container)
```
(venv) [root@I test-pvapy]# PVAPY_EPICS_LOG_LEVEL=0 python server.py | grep -v "UDP"
2023-09-28T17:50:32.468 Creating acceptor to 0.0.0.0:5075.
2023-09-28T17:50:32.468 Accepting connections at 0.0.0.0:5075.
2023-09-28T17:50:32.468 Creating datagram socket from: 0.0.0.0:49368.
2023-09-28T17:50:32.468 Broadcast address #0: 130.246.91.255:5076. (not unicast)
2023-09-28T17:50:32.468 Broadcast address #1: 192.168.122.255:5076. (not unicast)
2023-09-28T17:50:32.468 Broadcast address #2: 172.17.255.255:5076. (not unicast)
2023-09-28T17:50:32.468 Creating datagram socket from: 130.246.91.227:5076.
2023-09-28T17:50:32.468 Creating datagram socket from: 130.246.91.255:5076.
2023-09-28T17:50:32.468 Creating datagram socket from: 130.246.91.236:5076.
2023-09-28T17:50:32.468 Creating datagram socket from: 130.246.91.255:5076.
```

@mdavidsaver I had a question as I think I have some confusion here, does the PVAGateway also forward the UDP messages from the client network it monitors? I was assuming that since I had left  `EPICS_PVA_NAME_SERVERS` unset in my client code, it would not be using the PVAGateway. I don't think that can be the case, as I think as the network in our docker compose setup is isolated from my test host, so there is no way for them to communicate apart from the PVAGateway. 

In that case PVAGateway might still be a factor in the bug.

**Comment by sveseli** [comment]
Perhaps it would be easier if we remove gateway out of the picture for now, and you try running against pvapy server alone? Also, adding timestamps to client and correlating log entries with pvget requests might be helpful.

**Comment by mdavidsaver** [comment]
@jembishop   Either here, on https://github.com/mdavidsaver/p4p/issues/124, or by email if you prefer, please provide a complete description of the network topology between your test client and server.  All `EPICS_PVA*` or other relevant environment variables set on all involved nodes.  And attach any gateway config files.

> @mdavidsaver I had a question as I think I have some confusion here, does the PVAGateway also forward the UDP messages from the client network it monitors?  ...

What p4p.gw does is not as simple as forwarding.  While having all of the same settings, a gateway client side is configured independently of clients to the gateway server side.

> ... I was assuming that since I had left EPICS_PVA_NAME_SERVERS unset in my client code, it would not be using the PVAGateway. ...

This would seem to me to imply an unusual topology for a gateway.  Depending on IP routing rules, you might be able to set `EPICS_PVA_NAME_SERVERS` to the IP of the (not gateway) PVA server.  However, you __must also set `EPICS_PVA_AUTO_ADDR_LIST=NO`__ to disable UDP broadcast searching.

fyi. you can use `pvinfo` or `pvxinfo` to see the IP address of the server end of a TCP connection.  In the context of this discussion, this should either be the IP of a gateway, or of the server.

**Comment by jembishop** [comment]
Sorry for the delay in my reply. We have found the version of p4p pvagateway on our production server was not upgraded to the latest version which had the recent fix by @mdavidsaver. We are in a user cycle at the moment, so I can't touch the deployment of `p4p` on that server at the moment. I think it is best to wait until we have upgraded the version of `p4p` in our shutdown period, before I progress further as that may simply fix the issue. Thanks for your help, I will get back to you in a few weeks.
